frontend/src/GameReportForm.tsx

```diff
@@ -65,93 +65,263 @@ const cities: Stronghold[] = [
+        },
+    },
+    corruption: {
+        value: UNINITIALIZED,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    didFellowshipReachMordor: {
+        value: null,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    mordorTrack: {
+        value: UNINITIALIZED,
+        error: "",
+        validate: () => null,
+    },
+    initialEyes: {
+        value: UNINITIALIZED,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    wasAragornCrowned: {
+        value: null,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    aragornCrownedTurn: {
+        value: UNINITIALIZED,
+        error: "",
+        validate: () => null,
+    },
+    capturedStrongholds: {
+        value: [],
+        error: "",
+        validate: () => null,
+    },
+    interestRating: {
+        value: -1,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    comment: { value: "", error: "", validate: () => null },
 };
 
 type ValueOf<T> = T[keyof T];
 
 function GameReportForm() {
     const [formData, setFormData] = useState(initialFormData);
+    const [errorOnSubmit, setErrorOnSubmit] = useState<string | null>(null);
 
     const handleInputChange = <K extends keyof FormData>(field: K) => {
-        return (value: FormData[K]) =>
-            setFormData({ ...formData, [field]: value });
+        return (value: FormData[K]["value"]) =>
+            setFormData((prevData) => ({
+                ...prevData,
+                [field]: { ...prevData[field], value },
+            }));
+    };
+
+    const validateField = <K extends keyof FormData>(field: K) => {
+        return () => {
+            setFormData((prevData) => {
+                const error = prevData[field].validate();
+                return error
+                    ? {
+                          ...prevData,
+                          [field]: { ...prevData[field], error },
+                      }
+                    : prevData;
+            });
+        };
     };
 
+    const validateForm = () => {
+        const formError = detectFormErrors(formData);
+        if (formError) {
+            objectKeys(formData).forEach((field) => {
+                const fieldError = formData[field].validate();
+                if (fieldError) {
+                    setFormData((prevData) => ({
+                        ...prevData,
+                        [field]: { ...prevData[field], error: fieldError },
+                    }));
+                }
+            });
+            return formError;
+        }
+        return null;
+    };
+
+    const toPayload = (formData: FormData) =>
```

Comment:
This only does the one layer of cleaning, which is stripping out the frontend-only data; saving redundant field cleanup for a separate PR (it should be small I think)